text/0000-opentelemetry.md

```diff
@@ -0,0 +1,153 @@
+correlate automatically to the parent span.
+
+Developers may also create a trace before calling out to Fabric in their client code.
+The current trace information will be sent along with the message to the peer.
+
+Finally, developers may like to report metrics as part of chaincode execution. They can use OpenTelemetry counters,
+gauges and other metrics constructs to report metric data.
+The metric endpoint is configured when the chaincode is deployed. 
+In the case of external chaincode, the developers can set standard [OpenTelemetry variables](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-environment-variables.md)
+to point at an OpenTelemetry collector instance.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Message headers
+
+The trace information is passed in as additional data to the message, using a new field containing headers.
+Trace headers are filled in by the SDK and read by the chaincode execution framework.
+
+## Changes to the chaincode shims
+
+Each shim implementation should be modified to explicitly handle the additional message field to capture the trace ID passed in.
+If no parent trace information is present, the shim continues and creates a new trace with no parent relationship.
+
+As a demonstration, in this [PR](https://github.com/hyperledger/fabric-chaincode-java/pull/153/files#diff-db5cb40a9966f929cdf9963d606b8961248f94aac09ed150491fdc41e73d82b9R94), the Java shim implementation `call()` method which calls the chaincode execution changes to embed a try/finally statement
+that creates and ends the span once execution is finished. It also captures exceptions thrown during execution to set the status of the trace as an error.
+
+The Java chaincode shim also changes to implement the MetricsProvider interface to [provide metrics through OpenTelemetry](https://github.com/hyperledger/fabric-chaincode-java/pull/153/files#diff-d97166bc281ed3eec5bb83ccd164698c2ee3ae1bd5e8e4ab6a8ea1d86068e951R24).
+
+The same set of changes must be applied to the Go and Javascript shim implementations.
+
+## Changes to client SDKs
+
+Client SDKs automatically create a trace to capture the call to the chain.
+
+The trace is created with the kind Span.Client.
+
+If a current trace exists, the new client trace adds it as its parent.
+
+The trace and span ID must be sent to the chain as a message header.
+
+The SDK should use the standard environment variable environment to let users define how and if they want to report
+trace data to an endpoint of their choosing.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+OpenTelemetry has not reached 1.0. It has unequal maturation between metrics, traces and logs, and across languages.
+However, traces and metrics are well ahead of logs, and Java, Go and Javascript are well supported.
+
+The changes require a change in the structure of Protobuf messages, which may require a major upgrade of Fabric.
```

Comment:
I don't think we would need to modify the protobuf messages.  We should be able to use gRPC metadata for this.  Have a look at https://github.com/grpc-ecosystem/go-grpc-middleware/tree/master/tracing/opentracing .

Reply:
I missed that, I'll try it out and change the RFC.

Reply:
OK, I changed the Java reference implementation PR I provided along this RFC, and I'm able to test out and make sure this all works. I have pushed my changes to the PR and this RFC.